### PR TITLE
clp-s: Pack tables into a single file.

### PR DESCRIPTION
**Pull request by wraymo**
# References

# Description

This PR implements a basic table-packing mechanism. Instead of writing each table to a separate file, it writes all tables into a single file and adds a metadata file to record the offset of each table.

# Validation performed

+ Compressed a portion of [mongodb](https://zenodo.org/records/10516285) dataset with and without table-packing.
+ Decompressed the two archives, reordered decompressed logs and compared them.
+ Executed query `c: REPL` on both archives, reordered query results and compared them.

### CODE REVIEW

File: components/core/src/clp_s/ArchiveReader.cpp [hunk 1]
```diff
@@ -3,81 +3,187 @@
...
 
-    m_var_dict->read_new_entries();
-    m_log_dict->read_new_entries();
-    m_array_dict->read_new_entries();
+    m_table_file_reader.open(m_archive_path + "/table");
+    m_metadata_file_reader.open(m_archive_path + "/metadata");
+}
+
+void ArchiveReader::read_metadata() {
+    constexpr size_t cDecompressorFileReadBufferCapacity = 64 * 1024;  // 16 KB
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveReader.cpp:22]: 64KB? Same thing in a few other places.

File: components/core/src/clp_s/ArchiveReader.cpp [hunk 2]
```diff
@@ -3,81 +3,187 @@
...
+    m_metadata_decompressor.open(m_metadata_file_reader, cDecompressorFileReadBufferCapacity);
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveReader.cpp:23]: Since the pattern for all of these readX functions is basically
```
m_X_decompressor.open(X_file_reader)
// do reading
m_X_decompressor.close()
```
do we really need a ZstdDecompressor for every different read function?
Could we instead create a single `m_archive_decompressor` and reuse it for all of these different functions? It would probably require some changes to the ZstdDecompressor class so that it can re-use the same buffer each time it gets reopened, but it might be worth it.

Curious what you think? If you don't think its worth it I won't push for it.

File: components/core/src/clp_s/ArchiveReader.cpp [hunk 3]
```diff
@@ -3,81 +3,187 @@
...
-        );
-
-        schema_reader->load();
-        m_schema_id_to_reader[schema_id] = schema_reader;
+    auto schema_reader = std::make_unique<SchemaReader>(
+            m_schema_tree,
+            schema_id,
+            m_id_to_table_metadata[schema_id].num_messages
+    );
+    append_reader_columns(schema_reader, should_extract_timestamp);
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveReader.cpp:81]: It's sort of weird reading this code since it looks like we're creating a unique_ptr, handing it off to this append_reader_columns function, and also returning it a few lines later.

It might be worth at least adding a comment that the schema_reader unique_ptr is being passed by reference here.

File: components/core/src/clp_s/ArchiveWriter.hpp [hunk 4]
```diff
@@ -88,7 +88,10 @@ class ArchiveWriter {
     int m_compression_level{};
 
     std::shared_ptr<SchemaTree> m_schema_tree;
-    std::map<int32_t, SchemaWriter*> m_schema_id_to_writer;
+    std::map<int32_t, SchemaWriter*> m_id_to_schema_writer;
+
+    FileWriter m_table_file_writer, m_metadata_file_writer;
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveWriter.hpp:93]:
```suggestion
    FileWriter m_table_file_writer;
    FileWriter m_metadata_file_writer;
```
I think our style guide recommends one declaration per line.

File: components/core/src/clp_s/ArchiveReader.hpp [hunk 5]
```diff
@@ -54,18 +104,42 @@ class ArchiveReader {
      */
     void close();
 
+    /**
+     * @return The schema ids in the archive.
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveReader.hpp:108]: Probably worth a comment that this schema_ids vector also defines the order that tables should be read in to avoid seeking backwards.

File: components/core/src/clp_s/ArchiveReader.hpp [hunk 6]
```diff
@@ -54,18 +104,42 @@ class ArchiveReader {
...
 
     std::shared_ptr<SchemaTree> m_schema_tree;
-    ReaderUtils::SchemaMap m_schema_map;
-    std::map<int32_t, SchemaReader*> m_schema_id_to_reader;
+    std::shared_ptr<ReaderUtils::SchemaMap> m_schema_map;
+    std::vector<int32_t> m_schema_ids;
+    std::map<int32_t, SchemaReader::TableMetadata> m_id_to_table_metadata;
 
-    std::shared_ptr<TimestampDictionaryReader> m_timestamp_dict;
+    FileReader m_table_file_reader, m_metadata_file_reader;
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveReader.hpp:141]:
```suggestion
    FileReader m_table_file_reader;
    FileReader m_metadata_file_reader;
```

File: components/core/src/clp_s/ArchiveReader.hpp [hunk 7]
```diff
@@ -54,18 +104,42 @@ class ArchiveReader {
...
+    ZstdDecompressor m_table_decompressor, m_metadata_decompressor;
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveReader.hpp:142]:
```suggestion
    ZstdDecompressor m_table_decompressor;
    ZstdDecompressor m_metadata_decompressor;
```
Or switch to a single ZstdDecompressor per other comment

File: components/core/src/clp_s/SchemaWriter.hpp [hunk 8]
```diff
@@ -40,19 +40,17 @@ class SchemaWriter {
...
-    void store();
+    void store(ZstdCompressor& compressor);
 
     /**
      * Closes the schema writer.
      * @return the compressed size of the schema table in bytes
      */
     [[nodiscard]] size_t close();
 
+    uint64_t get_num_messages() const { return m_num_messages; }
```
Comment (gibber9809) [inline components/core/src/clp_s/SchemaWriter.hpp:51]: Missing docstring.

File: components/core/src/clp_s/ZstdDecompressor.cpp [hunk 9]
```diff
@@ -235,4 +235,5 @@ void ZstdDecompressor::reset_stream() {
 
     m_compressed_stream_block.pos = 0;
 }
+
```
Comment (gibber9809) [inline components/core/src/clp_s/ZstdDecompressor.cpp:238]:
```suggestion
```
Unnecessary whitespace

File: components/core/src/clp_s/ArchiveWriter.cpp [hunk 10]
```diff
@@ -49,13 +44,32 @@ size_t ArchiveWriter::close() {
     compressed_size += m_array_dict->close();
     compressed_size += m_timestamp_dict->close_local();
 
-    for (auto& i : m_schema_id_to_writer) {
-        i.second->store();
-        compressed_size += i.second->close();
+    m_table_file_writer.open(m_archive_path + "/table", FileWriter::OpenMode::CreateForWriting);
```
Comment (gibber9809) [inline components/core/src/clp_s/ArchiveWriter.cpp:47]: Could we change the filenames for the `table` and `metadata` sections?

I think it should at least be `tables` and `table_metadata`, but we could also consider using our "official" terminology and do something like `encoded_record_tables` or `erts` and `ert_metadata`.

File: components/core/src/clp_s/ArchiveReader.cpp (hunk 1 above)
Comment (wraymo) [reply components/core/src/clp_s/ArchiveReader.cpp:22]: Yeah, that's a mistake

File: components/core/src/clp_s/ArchiveReader.cpp (hunk 3 above)
Comment (wraymo) [reply components/core/src/clp_s/ArchiveReader.cpp:81]: Probably it's better to create a unique_ptr inside append_reader_columns and return it?

File: components/core/src/clp_s/SchemaWriter.hpp (hunk 8 above)
Comment (wraymo) [reply components/core/src/clp_s/SchemaWriter.hpp:51]: I remembered we don't need docstrings for basic getters/setters

File: components/core/src/clp_s/ArchiveWriter.cpp (hunk 10 above)
Comment (wraymo) [reply components/core/src/clp_s/ArchiveWriter.cpp:47]: Yeah, they are better

File: components/core/src/clp_s/SchemaWriter.hpp (hunk 8 above)
Comment (gibber9809) [reply components/core/src/clp_s/SchemaWriter.hpp:51]: Fair. Feel free to ignore this then.

File: components/core/src/clp_s/ArchiveReader.cpp (hunk 3 above)
Comment (gibber9809) [reply components/core/src/clp_s/ArchiveReader.cpp:81]: Yeah, that's a good idea. Maybe also rename it to `create_schema_reader` or something like that since it'll be doing all of the initialization.

File: components/core/src/clp_s/ArchiveReader.cpp (hunk 2 above)
Comment (gibber9809) [reply components/core/src/clp_s/ArchiveReader.cpp:23]: Yeah one decompressor object that we use for both, and we change the ZstdDecompressor implementation so that it can keep its internal buffer when it gets closed.

Comment (wraymo) [reply components/core/src/clp_s/ArchiveReader.cpp:23]: For the first case, I don't think it can reuse the buffer because it needs to read from a different file. For the second case, the buffer can be reused in two different ways. 1): create a memory mapped file and use `CompressedDataBuf` as `InputType` (what glt does) 2): create a method `Decompressor::try_seek_to_file_pos`. In this method, we call `FileReader::try_seek_to_pos` and reset the `m_compressed_stream_block`, but still use the same `m_file_read_buffer`. A potential optimization is to record `current_file_pos` after each  and see if we can use some data from `m_compressed_stream_block`. Which one do you prefer? And do you think I should add it in this PR or another PR?

Comment (gibber9809) [reply components/core/src/clp_s/ArchiveReader.cpp:23]: Potentially the first, but yeah we can leave it for another PR and talk about it more later.

**Review by gibber9809 — APPROVED** [review]
LGTM. PR title seems good as commit message too.

File: components/core/src/clp_s/SchemaReader.hpp [hunk 11]
```diff
@@ -81,7 +84,7 @@ class SchemaReader {
     /**
      * Loads the encoded messages
      */
-    void load();
+    void load(ZstdDecompressor& decompressor);
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/SchemaReader.hpp:87]: Update docstring

File: components/core/src/clp_s/SchemaWriter.hpp [hunk 12]
```diff
@@ -40,19 +40,17 @@ class SchemaWriter {
     /**
      * Stores the schema to disk.
      */
-    void store();
+    void store(ZstdCompressor& compressor);
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/SchemaWriter.hpp:43]: Update docstring.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
